### PR TITLE
Update to Typescript definition

### PR DESCRIPTION
**Pull request by DenisCarriere**
- [x] Extend class using `Key` & `Value` (defaults to `any`)
- [x] Add Typescript tests for `forEach`
- [x] Rename `fn` to `callback` in `forEach`
- [x] Add `index` to `forEach` callback
- [x] Added destroy() Typescript definition
- [x] Add Type tests, asserts the expected types (`test/types.ts`)

### CODE REVIEW

**Comment by DenisCarriere** [comment]
@w8r Let me know if you have any questions about this PR.

**Comment by w8r** [comment]
@DenisCarriere could you please add
```js
/**
 * @typedef {*} Key
 */
```
and 
```js
/**
 * @typedef {*} Value
 */
```
as a hommage to Google Closure compiler, in case I will at some point use it on this code?

**Comment by w8r** [comment]
And thanks a lot for your PR!

**Comment by DenisCarriere** [comment]
👍 That's pretty cool, never heard about `@typedef` before, looks pretty good for custom values.

**Comment by DenisCarriere** [comment]
Also going to add "Nullable types" to the JSDocs.

http://usejsdoc.org/tags-type.html

**Comment by w8r** [comment]
yes yes, pretty much *all* of it was in Google Closure lib and compiler since 2007 )) it's a pity it's such a complex thing to use

**Comment by DenisCarriere** [comment]
That's too bad 😢 Looks pretty solid.

There's also `@callback` in JSDocs, for the `comparator`.

http://usejsdoc.org/tags-callback.html

**Comment by w8r** [comment]
`@callback` I am not sure, first time I see it and there's nothing async in the library anyway so that might be misleading

**Comment by DenisCarriere** [comment]
You've got a few callbacks, it has nothing to do with async, here's how it looks:

It's just a bit cleaner to read and you can make your callbacks more descriptive.

## forEach
```js
  /**
   * Callback for forEach
   * @callback forEachCallback
   * @param {Node} node
   * @param {number} index
   */

  /**
   * @param  {forEachCallback} callback
   * @return {AVLTree}
   */
  forEach(callback) {
```
## Comparator
```js
  /**
   * Callback for comparator
   * @callback comparatorCallback
   * @param {Key} a
   * @param {Key} b
   * @returns {number}
   */

  /**
   * @class AVLTree
   * @constructor
   * @param  {comparatorCallback} [comparator]
   * @param  {boolean}            [noDuplicates=false] Disallow duplicates
   */
  constructor (comparator, noDuplicates = false) {
```

**Comment by DenisCarriere** [comment]
This is really cool! The `@typedef` preserves the `Key` type in VSCode (good job Microsoft!).

Might go and add some of these `@typedef` in TurfJS.

```js
/**
 * @typedef {*} Key
 */

/**
 * @typedef {*} Value
 */

/**
 * Default comparison function
 * @param {Key} a
 * @param {Key} b
 * @returns {number}
 */
function DEFAULT_COMPARE (a, b) { return a > b ? 1 : a < b ? -1 : 0; }
```

## `any`

image


## `number`
image

CC: @tmcw, @stebogit, @rowanwins

**Comment by w8r** [comment]
Could you please fix the `Node` `@typedef` to valid JSDoc? and I will merge immediately

**Comment by DenisCarriere** [comment]
I did?.. can you reference which lines you are talking about

File: src/index.js [hunk 1]
```diff
@@ -7,12 +7,12 @@ import { print, isBalanced } from './utils';
...
- *   balanceFactor: Number,
- *   key:           any,
- *   data:          object?
+ *   parent:        ?Node,
+ *   left:          ?Node,
+ *   right:         ?Node,
+ *   balanceFactor: number,
+ *   key:           Key,
+ *   data:          Value
  * }} Node
```
Comment (DenisCarriere) [inline src/index.js]: Isn't this correct?

**Comment by w8r** [comment]
You are right, nullable type. I used to document them explicitly. Thank you! I will merge and release

Kind regards,
Alexander Milevski

**Pull request merged**